src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -0,0 +1,572 @@
+/*
+ *    Copyright (c) 2024, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ *    This software is licensed under the Apache License, Version 2.0 (the
+ *    "License") as published by the Apache Software Foundation.
+ *
+ *    You may not use this file except in compliance with the License. You may
+ *    obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *    License for the specific language governing permissions and limitations
+ *    under the License.
+ */
+
+package io.supertokens.bulkimport;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+import io.supertokens.Main;
+import io.supertokens.bulkimport.exceptions.InvalidBulkImportDataException;
+import io.supertokens.config.CoreConfig;
+import io.supertokens.emailpassword.PasswordHashingUtils;
+import io.supertokens.emailpassword.exceptions.UnsupportedPasswordHashingFormatException;
+import io.supertokens.featureflag.EE_FEATURES;
+import io.supertokens.featureflag.FeatureFlag;
+import io.supertokens.multitenancy.Multitenancy;
+import io.supertokens.pluginInterface.Storage;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser.LoginMethod;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser.UserRole;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser.TotpDevice;
+import io.supertokens.pluginInterface.exceptions.StorageQueryException;
+import io.supertokens.pluginInterface.multitenancy.AppIdentifier;
+import io.supertokens.pluginInterface.multitenancy.TenantConfig;
+import io.supertokens.pluginInterface.multitenancy.TenantIdentifier;
+import io.supertokens.pluginInterface.multitenancy.exceptions.TenantOrAppNotFoundException;
+import io.supertokens.storageLayer.StorageLayer;
+import io.supertokens.utils.Utils;
+import io.supertokens.utils.JsonValidatorUtils.ValueType;
+
+import static io.supertokens.utils.JsonValidatorUtils.parseAndValidateFieldType;
+import static io.supertokens.utils.JsonValidatorUtils.validateJsonFieldType;
+
+public class BulkImportUserUtils {
+    private String[] allUserRoles;
+    private Set<String> allExternalUserIds;
+
+    public BulkImportUserUtils(String[] allUserRoles) {
+        this.allUserRoles = allUserRoles;
+        this.allExternalUserIds = new HashSet<>();
+    }
+
```

Comment:
in the bulkimport user type, we also need to allow plain text password cause it will be used during lazy migration task

Reply:
This is done.